### DISCUSSION TITLE
How do I add a cooldown to my command and how do I send a message if the user tries to use a command and the cooldown period isnt finished???

### DISCUSSION THREAD

**Discussion by samjesus8** [DSharpPlus/DSharpPlus#1390] [Support] [answered]
# Summary
I have this simple test command here and I have created a new cooldown based on the documentation. I tried searching long and hard on the documentation, Stack Overflow, and different youtube videos but none of them actually say how I can take the command timeout and send a message based on if the cooldown is active. I know I can add a cooldown when I'm declaring the Command names but I can't do anything with it since it's not a variable.

# Details
My main aim is simply to detect if the cooldown is active and I want to send an embed saying that the user cannot use the command until X amount of seconds has elapsed. It does not say in the documentation how I can even await the CoolDownAttribute after I have created it as a variable. I tried creating if statements based on the Remaining Cooldown but that didn't work either. Can someone please help me figure out how to get the remaining time of a cooldown and send a message based on that?

I want my outcome to be:

IF the cooldown period is over, users can use the command as normal
IF the cooldown period is not over, restrict the use of the command and send a message saying they cannot use the command and how much time is left till they can use it again

```
        [Command("test")]
        public async Task CooldownTest(CommandContext ctx) 
        {
            var cooldown = new CooldownAttribute(1, 10, CooldownBucketType.User);
        }
```

**Comment by Naamloos** [comment, marked as answer]
Hook into the `CommandErrored` event on your `CommandsNextExtension`, and check for the `Checks​FailedException`.

Along with that, you're using the cooldown attribute wrong. It's an attribute, use it just like `[Command("test")]`, 
e.g. `[Cooldown(1, 10, CooldownBucketType.User)]`

**Reply by samjesus8** [reply]
Ok so I've managed to do something that works. So far I've managed to get it to send a message when the exception is thrown

```
        private async Task OnCommandError(CommandsNextExtension sender, CommandErrorEventArgs e)
        {
            if (e.Exception is ChecksFailedException) 
            {
                Console.Error.WriteLine("Error " + e.Exception);

                var cooldownErrorMessage = new DiscordMessageBuilder()
                    .AddEmbed(new DiscordEmbedBuilder()
                    .WithTitle("You must wait for the cooldown to end")
                    );

                await e.Context.Channel.SendMessageAsync(cooldownErrorMessage);
            }
        }
```

So now I want to make this better by putting a description that says how many seconds of the cooldown is left. So how do I pull the time from the executed command and display it in the message so the user knows when the cooldown is finished???

**Reply by Giggitybyte** [reply]
See: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#declaration-and-type-patterns
